### ISSUE TITLE
❓ Get captured photo as blob 

### ISSUE THREAD

**Issue by ericlifs** [cuvent/react-native-vision-camera#737]
### Question

First thank you for this awesome library, all is working smoothly. I have a related question about the camera captured photos: is there a way of getting the photo content as a blob to easily send it to a server?

### What I tried

I have tried using several approaches like doing a `fetch(base64)` and then a `.blob()` but it's not working for me. I'm not sure if someone else has this same requirement and they were able to accomplish it.

### VisionCamera Version

2.11.2

### Additional information

- [X] I am using Expo
- [X] I have read the [Troubleshooting Guide](https://mrousavy.github.io/react-native-vision-camera/docs/guides/troubleshooting)
- [X] I agree to follow this project's [Code of Conduct](https://github.com/mrousavy/react-native-vision-camera/blob/main/CODE_OF_CONDUCT.md)
- [X] I searched for similar questions [in the issues page](https://github.com/mrousavy/react-native-vision-camera/issues?q=is%3Aopen+is%3Aissue+label%3A%22❓+question%22) as well as [in the discussions page](https://github.com/mrousavy/react-native-vision-camera/discussions) and found none.

**Comment by SeamusY** [comment]
I want to know too.... been stuck on that, and base64 is too large that it caused problems for uploading... I was wondering if their a more better compression method, to keep the quality decent but the file size smaller.

**Comment by SeamusY** [comment]
https://github.com/itinance/react-native-fs/issues/417

@ericlifs could try this, I am attempting that. Will update you on the attempt.

**Comment by mrousavy** [comment]
Hey! I decided to not expose a base64 string for `takePhoto` because that is insanely slow and clunky to get it transferred over the old Bridge. Instead, I am working on a way better solution: https://github.com/mrousavy/react-native-jsi-image

**Comment by JeftaJovanovic** [comment]
you can use fetch to get blob with the path you get from takePhoto.
`const getBlob = fileUri => {
  return fetch(fileUri).then(
    response => {
      return response.blob();
    },
    error => {
      console.log(`getBlob error`, error);
    },
  );
};`

**Comment by clovisrodriguez** [comment]
@JeftaJovanovic I have tried to do something like that with a video I recorded, but every time I tried to fetch the video store on the temp folder it fails, I'm using the URI I got back from the stop recording and the video renders perfect, I need this to create a blob that I could upload to our VOD provider.

the URI that I got back is: 

`file:///data/user/0/com.tbitmobile.qa/cache/VisionCamera-20220329_2144017934648742044078934.mp4`

and I have tried different combinations:

```
/data/user/0/com.tbitmobile.qa/cache/VisionCamera-20220329_2144017934648742044078934.mp4
file://data/user/0/com.tbitmobile.qa/cache/VisionCamera-20220329_2144017934648742044078934.mp4
data/user/0/com.tbitmobile.qa/cache/VisionCamera-20220329_2144017934648742044078934.mp4
```

none of this has worked as expected, my code
```
const onSubmit = async () => {
    try {
      const uri = video.path;
      console.log(uri);
      const file = await fetch(uri);
      console.log(file);
    } catch (e: any) {
      console.log(e.message);
      // Network request failed
    }
    // dispatch(uploadVideo(video.path))
  };
```

**Comment by mrousavy** [comment]
I don't think React Native's `fetch` allows you to fetch local files... 🤔

**Comment by clovisrodriguez** [comment]
Here is what I'm trying to achieve according to one of MUX video guides,
```
async function uploadVideo () {
  // videoUri here is the local URI to the video file on the device
  // this can be obtained with an ImagePicker library like expo-image-picker
  const imageResponse = await fetch(videoUri)
  const blob = await imageResponse.blob()

  // Create an authenticated Mux URL
  // this request should hit your backend and return a "url" in the
  // response body
  const uploadResponse = await fetch('/backend-api')
  const uploadUrl = (await uploadResponse.json()).url

  try {
    let res = await fetch(uploadUrl, {
      method: 'PUT',
      body: blob,
      headers: { "content-type": blob.type}
    });
    console.log("Upload is complete");
  } catch(error) {
    console.error(error);
  }
};
```

**Comment by clovisrodriguez** [comment]
https://docs.mux.com/guides/video/upload-files-directly#2-use-that-url-to-upload-in-your-client

**Comment by mrousavy** [comment]
Huh okay, maybe this does work. What exactly is the error message when you try to fetch a local file from VisionCamera?

**Comment by clovisrodriguez** [comment]
Well I tried to reproduce it today, but it worked! but I want to add my insight on what I did to solve it, from another StackOverflow thread of somebody else trying to access a resource inside the app cache it seems that didn't worked due to a lack of permissions, so I added a request for reading external storage on android, test on iOS is pending but thanks for replaying so soon.

**Comment by clovisrodriguez** [comment]
my response now looks like:
```
bodyUsed: false
headers: Headers {map: {…}}
ok: false
status: 0
statusText: ""
type: "default"
url: ""
_bodyBlob: Blob {_data: {…}}
_bodyInit: Blob {_data: {…}}
```

**Comment by og-abdul-mateen** [comment]
So, what's the proposed solution @mrousavy if I want to upload the image captured via this SDK to a server that expects a JPG or PNG file?

**Comment by elouisramsey** [comment]
> So, what's the proposed solution @mrousavy if I want to upload the image captured via this SDK to a server that expects a JPG or PNG file?

@mrousavy any help on this would be appreciated

**Comment by waleedshkt** [comment]
I strongly believe that base64-encoded `blob` should be added as a prop in `PhotoFile`. Because many times, there will be a need to upload the  photo to the server. 

Therefore, having a blob from the photo right away would greatly ease the uploading process. 

Right now, the best solution is to use either [react-native-fs](https://www.npmjs.com/package/react-native-fs) or [rn-fetch-blob](https://www.npmjs.com/package/rn-fetch-blob)

**Comment by mrousavy** [comment]
> I strongly believe that base64-encoded blob should be added as a prop in PhotoFile. Because many times, there will be a need to upload the photo to the server.
> 
> Therefore, having a blob from the photo right away would greatly ease the uploading process.

Great, yea, I'd be happy to review and accept a PR for this! :)

**Comment by akash-dey** [comment]
This solution worked!
TLDR: Convert `react-native-vision-camera` PhotoFile to blob using `fetch` and thereby convert blob to base64 using `FileReader`

``` 
const getBlob = useCallback((fileUri: string) => {
    return fetch(fileUri).then(
      response => {
        return response.blob()
      },
      error => {
        Logger.log(`Error in converting image to blob - ${error}`)
      },
    )
  }, [])

  const blobToBase64 = useCallback(blob => {
    try {
      const reader = new FileReader()
      reader.readAsDataURL(blob)
      return new Promise(resolve => {
        reader.onloadend = () => {
          resolve(reader.result)
        }
      })
    } catch (err) {
      Logger.log(`Error in converting blob to base64 - ${err}`)
      throw err
    }
  }, [])

  const sendImage = useCallback(
    async (imageUri: string) => {
      try {
        const blob = await getBlob(imageUri)
        blobToBase64(blob).then(finalImage =>
        // FinalImage is the finally converted base64 string
          webViewRef?.current?.injectJavaScript(
            `window.setImageCaptured('${finalImage}')`,
          ),
        )
      } catch (err) {
        Logger.log(err)
      }
    },
    [blobToBase64, getBlob],
  )
  ```

**Comment by tyhudson** [comment]
@mrousavy did we ever add this feature to vision camera?

**Comment by dtembo98** [comment]
> This solution worked! TLDR: Convert `react-native-vision-camera` PhotoFile to blob using `fetch` and thereby convert blob to base64 using `FileReader`
> 
> ```
> const getBlob = useCallback((fileUri: string) => {
>     return fetch(fileUri).then(
>       response => {
>         return response.blob()
>       },
>       error => {
>         Logger.log(`Error in converting image to blob - ${error}`)
>       },
>     )
>   }, [])
> 
>   const blobToBase64 = useCallback(blob => {
>     try {
>       const reader = new FileReader()
>       reader.readAsDataURL(blob)
>       return new Promise(resolve => {
>         reader.onloadend = () => {
>           resolve(reader.result)
>         }
>       })
>     } catch (err) {
>       Logger.log(`Error in converting blob to base64 - ${err}`)
>       throw err
>     }
>   }, [])
> 
>   const sendImage = useCallback(
>     async (imageUri: string) => {
>       try {
>         const blob = await getBlob(imageUri)
>         blobToBase64(blob).then(finalImage =>
>         // FinalImage is the finally converted base64 string
>           webViewRef?.current?.injectJavaScript(
>             `window.setImageCaptured('${finalImage}')`,
>           ),
>         )
>       } catch (err) {
>         Logger.log(err)
>       }
>     },
>     [blobToBase64, getBlob],
>   )
> ```

```
 const blob = await getBlob(imageUri)
```
Throws an error ```Error in converting image to blob - TypeError: Network request failed ```

**Comment by tyhudson** [comment]
Thank you for your response.  I solved it in my code by doing this. 

<img width="494" alt="image" src="https://user-images.githubusercontent.com/118188666/235684579-51b06d59-8be4-4c1c-9a52-96d292ccdffc.png">

<img width="463" alt="image" src="https://user-images.githubusercontent.com/118188666/235685723-9b9a3a76-f399-49c6-a2a3-05dd22291329.png">

I do like your solution too.  I thought it would have been awesome to pass a true/false to the image capture and get that back instead of saving it. 

It's all good everything is working.

**Comment by emadel1990** [comment]
> Thank you for your response. I solved it in my code by doing this.
> 
> <img alt="image" width="494" src="https://user-images.githubusercontent.com/118188666/235684579-51b06d59-8be4-4c1c-9a52-96d292ccdffc.png">
> 
> <img alt="image" width="463" src="https://user-images.githubusercontent.com/118188666/235685723-9b9a3a76-f399-49c6-a2a3-05dd22291329.png">
> 
> I do like your solution too. I thought it would have been awesome to pass a true/false to the image capture and get that back instead of saving it.
> 
> It's all good everything is working.

This work for me also! Thanks friend!

**Comment by emadel1990** [comment]
```
export const CustomCamera = () => {
  const cameraState = useAppSelector(getCameraState);
  const dispatch = useAppDispatch();
  const devices = useCameraDevices();
  const isAppForeground = useIsAppForeground();
  const device = devices.front || devices.back;
  const camera = useRef<Camera>(null);

  useEffect(() => {
    (async () => {
      Camera.requestCameraPermission();
    })();
    dispatch(setIsOpen(false));
  }, []);


  const takePhoto = async () => {
    if (!camera.current) return console.log('No camera');
    const photo = await camera.current?.takePhoto();
    dispatch(setIsOpen(false));
    fetchImage('file://${photo.path}');
  };

  const fetchImage = async (uri: string) => {
    const imageResponse = await fetch(uri);
    const imageBlob = await imageResponse.blob();
    const base64Data = await blobToBase64(imageBlob);
    dispatch(setPhoto(base64Data));
  };

  const blobToBase64 = (blob: any): Promise<string> => {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onerror = reject;
      reader.onload = () => {
        resolve(String(reader.result));
      };
      reader.readAsDataURL(blob);
    });
  };

  if (!device || device === null) return <Text>No camera</Text>;
  return (
    <>
      {cameraState.isOpen ? (
        <View style={styles.container}>
          <Camera
            isActive={isAppForeground}
            ref={camera}
            style={StyleSheet.absoluteFill}
            device={device}
            photo={true}
          />
          <Pressable onPress={takePhoto} style={styles.touchableOpacity} />
          <Pressable
            onPress={() => dispatch(setIsOpen(false))}
            style={styles.touchableOpacityClose}
          >
            <Icon style={{ zIndex: 2 }} fill={'#fff'} name="close-outline" />
          </Pressable>
        </View>
      ) : null}
    </>
  );
};
```

```
const styles = StyleSheet.create({
  touchableOpacityClose: {
    width: 35,
    height: 35,
    borderRadius: 30,
    backgroundColor: 'black',
    position: 'absolute',
    top: 15,
    right: 15,
    alignSelf: 'center',
    borderColor: '#f1f1f1',
    borderWidth: 3,
  },
  touchableOpacity: {
    width: 50,
    height: 50,
    borderRadius: 30,
    backgroundColor: 'grey',
    position: 'absolute',
    bottom: 25,
    alignSelf: 'center',
    borderColor: '#f1f1f1',
    borderWidth: 3,
  },
  container: {
    top: 0,
    left: 0,
    zIndex: 100,
    flex: 1,
    width: '100%',
    height: '100%',
    position: 'absolute',
  },
  item: {
    margin: 24,
    fontSize: 18,
    fontWeight: 'bold',
    textAlign: 'center',
  },
});
```


This is my component fully functional!

**Comment by tyhudson** [comment]
I also have code at the end of a process of taking a bunch of manual images, that deletes the cached files after they upload to a server.

**Comment by felkru** [comment]
Did anybody figure out a solution for Firebase cloud storage already? My code looks like this:
```javascript
const capturePhoto = async () => {
    const photo = await camera.current.takePhoto({
      flash: 'off',
    });
    setPath(photo.path);
    uploadFile(makeJsFile(photo.path));
};
```
I already implemented the uploadFile() function, but I wasn't able to get the makeJsFile() function work yet.

**Comment by felkru** [comment]
You dont need that. It's just as simple as:
```javascript
fetch(photo.path)
      .then(res => {
        uploadFile(res._bodyBlob, 'test.jpg');
      })
      .catch(err => {
        console.log('err', err);
      });
```

**Comment by mrousavy** [comment]
yep use fetch

**Comment by MateusHoffman** [comment]
@mrousavy @felkru  how to get the base64 of a frame that comes from useFrameProcessor?

**Comment by mrousavy** [comment]
frame.toArrayBuffer() - and please stop spamming this question.

There's no base64 for a GPU frame
